File: enade_time/frontend/src/api/benchmark.ts
import { apiClient } from "./client";
import type {
  BenchmarkComparativo,
  BenchmarkEtapa,
  BenchmarkExecucao,
  BenchmarkMetrica,
} from "@/types/api";

export async function fetchExecucoes(
  apenasValidas: boolean = true,
): Promise<BenchmarkExecucao[]> {
  const { data } = await apiClient.get<BenchmarkExecucao[]>(
    "/api/benchmark/execucoes",
    { params: { apenas_validas: apenasValidas } },
  );
  return data;
}

export async function fetchExecucao(
  id: number,
  apenasValidas: boolean = true,
): Promise<BenchmarkExecucao> {
  const { data } = await apiClient.get<BenchmarkExecucao>(
    `/api/benchmark/execucoes/${id}`,
    { params: { apenas_validas: apenasValidas } },
  );
  return data;
}

export async function fetchEtapas(
  execucaoId: number,
  apenasValidas: boolean = true,
): Promise<BenchmarkEtapa[]> {
  const { data } = await apiClient.get<BenchmarkEtapa[]>(
    `/api/benchmark/etapas/${execucaoId}`,
    { params: { apenas_validas: apenasValidas } },
  );
  return data;
}

export async function fetchMetricas(
  apenasValidas: boolean = true,
): Promise<BenchmarkMetrica[]> {
  const { data } = await apiClient.get<BenchmarkMetrica[]>(
    "/api/benchmark/metricas",
    { params: { apenas_validas: apenasValidas } },
  );
  return data;
}

export async function fetchComparativo(
  apenasValidas: boolean = true,
): Promise<BenchmarkComparativo> {
  const { data } = await apiClient.get<BenchmarkComparativo>(
    "/api/benchmark/comparativo",
    { params: { apenas_validas: apenasValidas } },
  );
  return data;
}

File: enade_time/api/routers/benchmark.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from psycopg2.extensions import cursor as PgCursor

from enade_time.api.dependencies import get_db_cursor
from enade_time.api.repositories import benchmark_repo
from enade_time.api.schemas.benchmark import (
    BenchmarkComparativo,
    BenchmarkEtapa,
    BenchmarkExecucao,
    BenchmarkMetrica,
    Campanha,
)
from enade_time.api.settings import Settings, get_settings

router = APIRouter(prefix="/benchmark", tags=["benchmark"])


def _uuid_str(valor: UUID | None) -> str | None:
    return str(valor) if valor is not None else None


@router.get("/execucoes", response_model=list[BenchmarkExecucao])
def execucoes(
    apenas_validas: bool = Query(True),
    campanha_id: UUID | None = Query(None, description="Restringe a uma campanha"),
    cursor: PgCursor = Depends(get_db_cursor),
    settings: Settings = Depends(get_settings),
) -> list[dict]:
    return benchmark_repo.listar_execucoes(
        cursor,
        apenas_validas=apenas_validas,
        ids_excluir=settings.benchmark_ids_excluir,
        campanha_id=_uuid_str(campanha_id),
    )


@router.get("/execucoes/{execucao_id}", response_model=BenchmarkExecucao)
def execucao_por_id(
    execucao_id: int,
    apenas_validas: bool = Query(True),
    cursor: PgCursor = Depends(get_db_cursor),
    settings: Settings = Depends(get_settings),
) -> dict:
    row = benchmark_repo.get_execucao(
        cursor,
        execucao_id=execucao_id,
        apenas_validas=apenas_validas,
        ids_excluir=settings.benchmark_ids_excluir,
    )
    if row is None:
        raise HTTPException(status_code=404, detail="Execução não encontrada")
    return row


@router.get("/etapas/{execucao_id}", response_model=list[BenchmarkEtapa])
def etapas(
    execucao_id: int,
    apenas_validas: bool = Query(True),
    cursor: PgCursor = Depends(get_db_cursor),
    settings: Settings = Depends(get_settings),
) -> list[dict]:
    rows = benchmark_repo.listar_etapas(
        cursor,
        execucao_id=execucao_id,
        apenas_validas=apenas_validas,
        ids_excluir=settings.benchmark_ids_excluir,
    )
    if rows is None:
        raise HTTPException(status_code=404, detail="Execução não encontrada")
    return rows


@router.get("/metricas", response_model=list[BenchmarkMetrica])
def metricas(
    apenas_validas: bool = Query(True),
    campanha_id: UUID | None = Query(None, description="Restringe a uma campanha"),
    cursor: PgCursor = Depends(get_db_cursor),
    settings: Settings = Depends(get_settings),
) -> list[dict]:
    return benchmark_repo.listar_metricas(
        cursor,
        apenas_validas=apenas_validas,
        ids_excluir=settings.benchmark_ids_excluir,
        campanha_id=_uuid_str(campanha_id),
    )


@router.get("/campanhas", response_model=list[Campanha])
def campanhas(cursor: PgCursor = Depends(get_db_cursor)) -> list[dict]:
    """Campanhas de benchmark (grupos de suítes), da mais recente para a mais antiga."""
    return benchmark_repo.listar_campanhas(cursor)


@router.get("/comparativo", response_model=BenchmarkComparativo)
def comparativo(
    apenas_validas: bool = Query(True),
    campanha_id: UUID | None = Query(
        None, description="Campanha a comparar (padrão: a oficial mais recente)"),
    suite_id: UUID | None = Query(None, description="Drill-down em uma suíte"),
    cursor: PgCursor = Depends(get_db_cursor),
    settings: Settings = Depends(get_settings),
) -> dict:
    return benchmark_repo.comparativo(
        cursor,
        apenas_validas=apenas_validas,
        ids_excluir=settings.benchmark_ids_excluir,
        campanha_id=_uuid_str(campanha_id),
        suite_id=_uuid_str(suite_id),
    )
